### ISSUE TITLE
Some MongoDB docs does not sync to Elasticsearch index

### ISSUE THREAD

**Issue by ail-13** [rwynn/monstache#272]
Hello, I have a problem with sync. Some changes in MongoDB are not sync in the Elasticsearch index. Right after `direct-read-namespaces`, the number of docs in Elasticsearch equal the number of MongoDB docs. Something around ~50k documents are created every day in MongoDB. And after a few days, the index may be missing ~100 documents in different collections. And the contents of some documents in the Elasticsearch index may be different from the documents in MongoDB. I can't find a pattern of this problem. Sometimes it happens when I insert 500 docs one by one, sometimes when I insert only several docs (5-10). What could be the reason? I use Mongodb 3.6.3 Elasticsearch 5.6.13 and Moustache v3.24.6. and the default settings for Monstache.

```
change-stream-namespaces = [
    "db.Modify",
    "db.User",
    ...
]
routing-namespaces = [
    "db.Modify",
    ...
]
direct-read-namespaces = [
    "db.ModifyView",
    "db.User",
    ...
]

[[mapping]]
namespace = "db.Modify"
index = "db"
[[mapping]]
namespace = "db.ModifyView"
index = "db"
type = "Modify"

[[relate]]
namespace = "db.Modify"
with-namespace = "db.ModifyView"
keep-src = false
[[relate]]
namespace = "db.Group"
with-namespace = "db.Modify"
src-field = "_id"
match-field = "groupId"
keep-src = false

[[script]]
namespace = "db.ModifyView"
routing = true
script = """
module.exports = function(doc) {
    doc._meta_monstache = { parent: doc.OrderId };
    return doc;
}
"""

[[mapping]]
namespace = "db.User"
index = "db"

...
```

**Comment by sbnajardhane** [comment]
@rwynn @ail-13 
I am also facing the same issue. [`here`](https://github.com/rwynn/monstache/issues/263)
Few records fail to sync to elasticsearch index.
When I have updated bulk documents in mongo (around 100k), around ~50 records where out of sync.
I have checked the trace.log of monstache, no entry where present for those 50 users in the trace log. However, the data correctly got synced into the mongo-secondary machine that means oplog is properly generated.
I am not able to figure out the cause of the issue, can anyone help me out here

**Comment by ail-13** [comment]
Does this issue occur in all versions of Moustache? Or on v4, v5, v6 there is no such problem?

**Comment by sbnajardhane** [comment]
I am using monstache 4.11.3, elasticsearch 6.3 and mongo 3.2
not sure about the other versions.

**Comment by ail-13** [comment]
I will get the same problem if I use version v 4.11.3. But about versions v5 and v6 it is not yet known. Did I understand you correctly?

**Comment by rwynn** [comment]
@sbnajardhane you can try latest monstache v5 with the instructions here for using against a pre-3.6 version of MongoDB.  

https://rwynn.github.io/monstache-site/advanced/

Note you will need to ensure `enable-oplog` is set to true.

**Comment by rwynn** [comment]
@ail-13 I'm not sure you can try latest v5 because you are not on Elasticsearch 6 yet.  You may get some issues with bulk requests.  

If you do have issues with it: you could either upgrade to Elasticsearch 6 (I think 5 might be deprecated soon).  Or you could checkout the monstache rel5 branch and update imports and go.mod to reference `gopkg.in/olivere/elastic.v5`.

**Comment by sbnajardhane** [comment]
> @sbnajardhane you can try latest monstache v5 with the instructions here for using against a pre-3.6 version of MongoDB.
> 
> https://rwynn.github.io/monstache-site/advanced/
> 
> Note you will need to ensure `enable-oplog` is set to true.

Will try out the version 5.

**Comment by ail-13** [comment]
It looks like my problem's solved. First I found a problem with my application, some operations with MongoDB exceeded the value of `relate-buffer` in Monstache and because of this, some data was not updated in the Elasticsearch index. I fixed it, but the problem was not solved completely, sometimes it could be repeated. After that, I updated MongoDB to v4.2.0, Elasticsearch to v6.6.1 and used Moustache v5.1.0. And now almost 2 weeks there are no problems with the index. Maybe it will be useful for someone
